### ISSUE TITLE
Add support to emit dynamic assemblies using ModuleBuilder

### ISSUE THREAD

**Issue by glopesdev** [dotnet/roslyn#10881]
As I understand it from [this stack overflow question](http://stackoverflow.com/questions/22974473/using-roslyn-emit-method-with-a-modulebuilder-instead-of-a-memorystream), in the past there was some experimental support to emit a compilation object into a `ModuleBuilder`. It has since disappeared from the Roslyn API.

It seems that `ModuleBuilder` and the rest of the `System.Reflection.Emit` namespace would be highly desirable targets for Roslyn, as they enable support for collectible assemblies and dynamic methods, which seem ideal for scripting environments.

Any plans for bringing this API back?

If not, would it still be possible to build such a translation layer on top of Roslyn syntax trees? How hard would it be (i.e. can you provide a description of how this feature is "problematic")?.

**Comment by per-samuelsson** [comment]
@glopesdev 

> If not, would it still be possible to build such a translation layer on top of Roslyn syntax trees? How hard would it be (i.e. can you provide a description of how this feature is "problematic")?.

It's not 1-1 with what you are asking, but maybe the future initiative of supporting *source generators* could interest you? It's originally proposed here:

https://github.com/dotnet/roslyn/issues/5561

Here's some live documentation:
https://github.com/dotnet/roslyn/blob/features/generators/docs/features/generators.md

And here's the branch if you want to really get into it:
https://github.com/dotnet/roslyn/tree/features/generators

**Comment by ManishJayaswal** [comment]
FYI @tmat @jaredpar

**Comment by svick** [comment]
@per-samuelsson I don't see how is that related at all. This feature request is about compiling C# code at runtime, generators are about generating C# code at compile time.

**Comment by tmat** [comment]
There are no plans to bring Reflection.Emit backend back to Roslyn. When we prototyped it a while ago we hit many issues with the API. Simply put Reflection.Emit was not designed to be used by a compiler backend. Collectible assemblies themselves have some limitations that prevented us to use them as containers for code generated by Roslyn scripting runtime. 

That said we haven't given up on collectible code generation, but it would take a new CLR feature to make it happen. Support for collectible code gen is tracked by the following UserVoice request:

https://visualstudio.uservoice.com/forums/121579-visual-studio-2015/suggestions/6120992-support-for-collectible-assemblies

Please cast your vote if you are interested.

**Comment by glopesdev** [comment]
@tmat Thank you for your feedback.

Could you give some examples of the problems of `Reflection.Emit`? I've seen in several places Roslyn developers hinting that this integration is "problematic", but I still don't understand exactly where it falls short, and it would be very helpful to have some pointers.

So your opinion is that there is no point in trying to take Roslyn generated syntax trees and converting them into `Reflection.Emit` modules and methods?

**Comment by per-samuelsson** [comment]
> @per-samuelsson I don't see how is that related at all. This feature request is about compiling C# code at runtime, generators are about generating C# code at compile time.

Yes, I understood that, and that's why I wrote it was not 1-1. I just figured that since source generators is something new, and both technologies is still about generated code (even though one is runtime IL, and the other is compile-time source), it might be of interest. Simply just trying to be creative and supportive, if it would turn out `Reflection.Emit` would in fact be dropped. Not forcing anything on no one.
